expression/scalar_function.go

```diff
@@ -125,6 +125,40 @@ func (sf *ScalarFunction) MarshalJSON() ([]byte, error) {
 	return []byte(fmt.Sprintf("\"%s\"", sf)), nil
 }
 
+// typeInferForNull infers the NULL datums field type and set the field type
+// of NULL datum same as other non-null operands.
+func typeInferForNull(args []Expression) {
+	if len(args) < 2 {
+		return
+	}
+	var isNull = func(expr Expression) bool {
+		cons, ok := expr.(*Constant)
+		return ok && cons.RetType.Tp == mysql.TypeNull && cons.Value.IsNull()
+	}
+	// Inference the actual field type of the NULL datum.
+	var retFieldTp *types.FieldType
+	var hasNullArg bool
+	for _, arg := range args {
+		isNullArg := isNull(arg)
+		if !isNullArg && retFieldTp == nil {
+			retFieldTp = arg.GetType()
+		}
+		hasNullArg = hasNullArg || isNullArg
+		// Break if there are both NULL and non-NULL expression
+		if hasNullArg && retFieldTp != nil {
+			break
+		}
+	}
+	if !hasNullArg || retFieldTp == nil {
+		return
+	}
+	for _, arg := range args {
+		if isNull(arg) {
```

Comment:
Can we reuse the result in line 142 instead of calculating again?

Reply:
We must allocate an array to store the result of `isNull`  if we want to reuse them, I think this calculation is faster than memory allocation in most scenario.